src/audio/chain_dma.c

```diff
@@ -224,6 +236,45 @@ static size_t chain_get_transferred_data_size(const uint32_t out_read_pos, const
 	return buff_size - in_read_pos + out_read_pos;
 }
 
+#if CONFIG_IPC_MAJOR_4
+static void xrun_notif_msg_init(struct ipc_msg *msg_xrun, uint32_t resource_id, uint32_t event_type)
+{
+	struct ipc4_resource_event_data_notification *notif_data = msg_xrun->tx_data;
+
+	notif_data->resource_id = resource_id;
+	notif_data->event_type = event_type;
+	notif_data->resource_type = GATEWAY;
+	notif_data->dws[0] = 0;
+	notif_data->dws[1] = 0;
+	notif_data->dws[2] = 0;
+	notif_data->dws[3] = 0;
+}
+
+static void handle_xrun(struct chain_dma_data *cd)
+{
+	if ((cd->link_connector_node_id.f.dma_type == ipc4_hda_link_output_class)
+	    && (!cd->under_over_run_notification_sent)) {
+		tr_warn(&chain_dma_tr, "handle_xrun(): underrun detected");
+		xrun_notif_msg_init(cd->msg_xrun, cd->link_connector_node_id.dw,
+				    SOF_IPC4_GATEWAY_UNDERRUN_DETECTED);
+		ipc_msg_send(cd->msg_xrun, NULL, true);
+		cd->under_over_run_notification_sent = true;
```

Comment:
This probably should be good to put behind a Kconfig. At least with current Linux SOF driver, I don't see what we could do with these messages. Is driver supposed to react to these, or are these solely for debugging?

Reply:
We should treat it as a useful information for debugging. Xrun notification message should be possible to be masked by driver (even in production fw), however we do not have this functionality right now in converge. In such case I will create new define in Kconfig for blocking xrun notifications.

Reply:
I suggest move this function to some file under ipc/ipc4

Reply:
@softwarecki I agree that 'xrun_notif_msg_init' could be used as generic function and it would be reasonable to move it to ipc/ipc4, however, I am not sure if 'handle_xrun' is good candidate as it is specific handling dedicated for chain dma interface.